packages/block-editor/src/hooks/color.js

```diff
@@ -293,6 +294,9 @@ export function ColorEdit( props ) {
 	const { clientId, name, attributes, setAttributes } = props;
 	const settings = useBlockSettings( name );
 	const isEnabled = useHasColorPanel( settings );
+
+	const [ userValue ] = useGlobalStyle( '', props.name, 'user' );
```

Comment:
I do think we need some kind of `useGlobalStyle` hook in the block editor or maybe just `useStyle` since we already have the `useSetting` one for the "settings". 

So yeah I wouldn't use the "global styles provider" and `useGlobalStyle` or `useGlobalSettings` as they exist today. In fact these two hooks were in edit-site as well, they have been moved to block-editor but as you can check they're only used in edit-site, so while I believe the goal was to "merge them" with "useSetting" and a potential "useStyle" that is block editor specific, this was never done. So at the moment, I consider "useGlobalStyle" and "useSetting" as edit-site specific things just like the provider. (I think it was @noisysocks that moved them, I forgot the entire reasoning)

I guess in my mind, we need to introduce a new key to settings prop of BlockEditorProvider that should contain the global styles "styles" property (we only have settings right now) and use that property to build a block-editor version of `useGlobalStyle` hook?

Maybe I'm missing something here so would love to hear more thoughts.

Reply:
> we need to introduce a new key to settings prop of BlockEditorProvider that should contain the global styles "styles" property (we only have settings right now) and use that property to build a block-editor version of useGlobalStyle hook?

Yes! This would certainly fulfill the needs of the current PR (and the larger styles inheritance UI system). I don't know enough to know if this potentially introduces any problems though.

Btw, is there any overlap here with https://github.com/WordPress/gutenberg/pull/55970 ?

Reply:
@michalczaplinski not necessarily, the problem here is a bit different than sharing the same code between edit-post and edit-site, we want access to the style hierarchy in the "block-editor" package. Until now, the block editor was only aware of the editor styles as CSS (`styles` property in the settings prop) but now we want the theme.json structure.

Reply:
I dug here a little bit and these are the ideas I tried:

#### 1. Create a `useStyles()` hook in the `block-editor` to get the styles and pass them to the `BlockEditorProvider`.
 
The `GlobalStylesProvider` that I used in the current PR relies on `useGlobalStylesUserConfig()`:

https://github.com/WordPress/gutenberg/blob/40a4b4bbf419e29c90857563555b19dab389c11f/packages/edit-site/src/components/global-styles/global-styles-provider.js#L33-L34

which in turn loads the global styles from the `core-data` store: 

https://github.com/WordPress/gutenberg/blob/40a4b4bbf419e29c90857563555b19dab389c11f/packages/edit-site/src/components/global-styles/global-styles-provider.js#L40-L46

However, we cannot use the `core-data` package in the `block-editor` either...

#### 2. Pass the data to the settings on the server

My second thought was to see if we can pass the required styles on the server in PHP in the [`block-editor-settings.php`](https://github.com/WordPress/gutenberg/blob/7bac5ee61413f4578757c56643dbe6afc548a8c2/lib/block-editor-settings.php). However, the Global Styles are loaded via the REST API so that's not an option either as far as I can tell.

#### 3. Create a `useStyles()` hook in the `editor` package and use it in the `edit-post` to inject the styles.

This approach looks the most reasonable to me as we can use `core-data` in the `editor` package. I opened a new PR in https://github.com/WordPress/gutenberg/pull/59929 following this approach.

However, I'm still not sure if that's the best way to solve the problem: 

1. Having distinct logic for getting the styles in the `edit-site` and `edit-post` does not seem ideal.
2. It seems a bit out of place to have styles-specific logic in [edit-post/src/editor.js](https://github.com/WordPress/gutenberg/blob/a3ff3be245fe4912c6715990152b5f563e096dcf/packages/edit-post/src/editor.js)

I'd appreciate your guidance here @youknowriad! 🙏

Reply:
For me, the best approach is probably an additional "setting" to the settings prop of the `BlockEditorProvider`. 

 - We have `__experimentalFeatures` (which we should probably rename at some point for the "settings" part of theme.json
 - We need the same but for "styles".

The problem is going to be the naming of these settings and I'm not the best person when it comes to naming things tbh 😬. cc @mcsf maybe

Reply:
it is also possible to combine both "__experimentalFeatures" and styles in a single prop we could name `theme` maybe but I'm not sure entirely. I'd defer to others.

Reply:
> an additional "setting" to the settings prop of the BlockEditorProvider.

Isn't that what I proposed in https://github.com/WordPress/gutenberg/pull/59929/ ? and referred to above in **3.** ?

I just called it `__globalStyles` for now: 

https://github.com/WordPress/gutenberg/pull/59929/files#diff-d00cfe048041886ee9b16f6f42422317b6e770cbea20cbe0e4d0ca7c8e0ec0e1R81

Reply:
@michalczaplinski you're right, sorry I missed this.

Reply:
No worries, I appreciate the guidance! I'm glad that we had roughly the same idea 🙂 We can continue to discuss in the PR.

Reply:
> it is also possible to combine both "__experimentalFeatures" and styles in a single prop we could name `theme` maybe but I'm not sure entirely. I'd defer to others.

How about _template_? :trollface: 

More seriously, I think this should be an opportunity to inventory everything that goes into `__experimentalFeatures` (as someone who isn't touching these pieces all the time, it's hard to remember what goes where). I think I'd favour a solution that strives for consistency with theme.json, since theme.json is already established as the dominant mental model that people need to get comfortable with, so we might as well piggyback off that.

Reply:
(Since the question was about naming, this was implicit but omitted in my comment: it would be nice to be able to express things in terms of `styles` & `settings`)

Reply:
> How about template? :trollface:
I'll do you one better: Template hooks 😄 

----

Makes sense @mcsf. I've closed this PR in favour of https://github.com/WordPress/gutenberg/pull/59929 which is the approach @youknowriad suggested. Let's continue to discuss there 🙂 .